### DISCUSSION TITLE
@types/node

### DISCUSSION THREAD

**Discussion by mcollina** [DefinitelyTyped/DefinitelyTyped#62117] [Issues with a @types package]
In the `@types/node` the [Legacy APIs](https://nodejs.org/api/documentation.html#stability-index) are flagged as `@deprecated`. Those APIs are widely used in the ecosystem and tutorials, and users are confused by them. Specifically, there is no good substitution for the `querystring` module as it's more than an order of magnitude faster than `URLSearchParam`.

Users call the APIs that Node.js flags as "legacy" as "deprecated", creating a significant confusion  for them. Those APIs are not deprecated and they will _not_ be removed. If somebody want to use them , they can.

**Comment by voxpelli** [comment]
Relevant lines: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/94472a5a7daaad59de1daba77750e4ab975a52fb/types/node/scripts/generate-docs/ast-processing.ts#L180-L190

I guess a PR would simply need to remove the `case Stability.Legacy:` and the `@deprecated` tags will be removed.

@mcollina Any reason why this generation and its script doesn't move to the Node.js team? And/or why the Node.js team doesn't publish its own types, maybe as eg. `@nodejs/types` on npm?

**Reply by voxpelli** [reply]
For wider context, here's a related Twitter thread: https://twitter.com/matteocollina/status/1566813876892704768

**Reply by voxpelli** [reply]
Sadly neither JSDoc (https://jsdoc.app/tags-deprecated.html) or TypeScript (https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated) define the any wider meaning for `@deprecated`

TypeScript says it will likely be rendered as a ~~strikethrough~~, but apart from that neither qualify whether there's an expectancy that the API will eventually be removed or not.

My interpretation is that the `@deprecated` description is meant to convey the nature of the deprecation, but for Node.js Legacy API:s that description simply says `Legacy`: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b7fe9e34b60493e51fced64c808cb3281bdf0a50/types/node/querystring.d.ts#L11

An alternative improvement would be to extend it with the full description + a link, eg:

```javascript
/**
 * @deprecated [Legacy](https://nodejs.org/api/documentation.html#stability-index). Although this feature is unlikely to be removed and is still covered by semantic versioning guarantees, it is no longer actively maintained, and other alternatives are available.
 */
```

It would then look like this, as can be seen in [this playground](https://www.typescriptlang.org/play?#code/PQKhCgAIUgBATApgBwE6IMYEMAuj6QDaAMogOZYYCeAugBQAWOOyAzgFzDAB2A9kgCtWAOl6oywLMgCWweLwwBXALaJuOXNN7dhTZQBsAxKw0Ajafuk4qAWmnckADwCUwyAEF9OBr0VkGkN7SrJAAZoi4iuiQwZCK3JYA1oj6VIG8kKaIkOjKvABu+JBYDjEhJhb6kBgFiOgEpmmsiMolONIYkIWorFrc9mSQZIpYqG2IiKwANDE4ZZB8kPraZHXFGO2FqZCt9hr2+DMlBLzea1heddyahSGj2Vj5WBZYpvqIwlAgwODWyNmhXgZAC8kBMqAGAG5wL8qP9MqNIKDAbxoeAgA):

![Skärmavbild 2022-09-05 kl  18 29 54](https://user-images.githubusercontent.com/34457/188489638-82afb435-798a-4cef-a86b-d5d5e64cf612.png)

**Reply by mcollina** [reply]
That's exactly what I mean, this is very confusing for people.
See https://github.com/fastify/fastify/issues/4239 and https://github.com/fastify/fastify/issues/3645

**Reply by voxpelli** [reply]
@mcollina Not sure those two are caused by the types necessarily though. Eg the Node.js docs themselves say actually say exactly that (emphasis mine):

> The querystring API is considered Legacy. While it is still maintained, **new code should use the `URLSearchParams` API instead**.

Probably should add at least a `likely` there so it becomes `should likely use` or dive deeper into the current speed differences?

**Reply by anonrig** [reply]
@voxpelli In terms of performance, for inputs that requires decoding, URLSearchParams are faster than `fast-querystring` by **~66%**, but for all other inputs, `fast-querystring` is actually faster than URLSearchParams (by **~166%**) and `node:querystring` (by **~10%**).

This is mainly related to JavaScript being really bad at decoding and encoding strings, and that C++ is really good at it.

<img width="477" alt="image" src="https://user-images.githubusercontent.com/1935246/188495039-80274917-68a4-4411-aa0e-4ef28076ecc3.png">

The only difference I could find between the URLSearchParams and alternative javascript solutions is that URLSearchParams omit leading `?` in the input, and others don't.

PS: All of those benchmarks can be found from https://github.com/anonrig/fast-querystring/blob/main/test/benchmark.mjs

**Reply by voxpelli** [reply]
@anonrig Great! Whether the Node.JS docs should be updated to mention this feels out of scope of this discussion though. Maybe an issue in the Node.js project should be opened for that? So this issue can get attacked from multiple angles 👍
